src/core/builder.rs

```diff
@@ -0,0 +1,354 @@
+                            &module_name
+                        );
+
+                        // If this is a package we need to process the names b/c they may be submodules
+                        let module_spec: Option<ModuleSpec> = find_spec(
+                            &module_name
+                        );
+
+                        if let Some(s) = module_spec {
+                            if s.is_package() {
+                                for alias in names {
+                                    self._process_dependency(
+                                        Some(&spec.name),
+                                        &format!(
+                                            "{}.{}",
+                                            module_name,
+                                            alias.name
+                                        )
+                                    )
+                                }
+                            }
+                        }
+                    } else {
+                        panic!("Broken assumption of implementation, revist this block to see if there are issues.")
+                    }
+                }
+            }
+        }
+        self.processing.remove(&spec.name);
+        debug!("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
+        debug!("Done '{}'", spec.name);
+        debug!("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
+    }
+
+    pub fn _process_dependency(&mut self, from: Option<&String>, name: &str) {
+        // Maybe expensive but some values will change names after find_spec()
+        // TODO: Deal with this in another way?
+        let spec: Option<ModuleSpec> = find_spec(name);
+
+        if spec.is_none() {
+            debug!("!!!! Unable to find spec for name: '{}' !!!!", name);
+            return
+        }
+
+        // Rebind spec & name to make things easier going forward
+        let spec = spec.unwrap();
+        let name = spec.name.clone();
+
+        // Double borrow in add_dependency will fail, self references are valid python, just going to not track them for now.
+        // TODO: Example??
+        if let Some(from) = from {
+            if from == &name {
+                return
+            }
+        }
+
+        self.metadata.processed += 1;
+        if let Some(from) = from {
+            debug!("Processing dependency: {} -> {}", from, name);
+        } else {
+            debug!("Processing dependency on: {}", name);
+        }
+
+
+        if self.graph.has_node(&name) {
+            // Just need to update the dependencies if required
+            if let Some(from) = from {
+                self.graph.add_dependency(from, &name)
+            }
+
+            // Done!
+            return
+        } else if name != "<terminal>" {
+            // TODO: Can this happen before reaching out to python
+            if !self.cache.is_none() {
+                if self.cache.as_ref().unwrap().has_node(&name) {
```

Comment:
Can remove a level of nesting and an unwrap if you add a function on `self.cache` to take an `Option<&str>`